### DISCUSSION TITLE
How to grow partitions using `systemd-growfs`?

### DISCUSSION THREAD

**Discussion by paulmenzel** [systemd/mkosi#2690] [Q&A]
Dear mkosi folks, as with `systemd-repart` I assume, I need to configure the image myself, if I want this feature. Instead of `/etc/fstab` using a [partition flag bit 59](https://uapi-group.org/specifications/specs/discoverable_partitions_specification/#partition-attribute-flags) can be used. Did somebody already integrate that in their mkosi configuration?

**Comment by rgilton** [comment]
The method I tried was was to put this file in `/etc/repart.d/50-root.conf`:

```
[Partition]
Type=root
GrowFileSystem=yes
```

I was expecting `systemd-repart` and `systemd-growfs` to do their magic and grow the root partition and filesystem.  Unfortunately neither of those things happened.  I think this is because the "first boot" is occurring when mkosi builds the image (i.e. it creates `/etc/machine-id`, which is what systemd uses to determine if the first boot has occurred), rather than when I ship the resulting image off somewhere else (in my case, libvirt).

If I manually start `systemd-growfs-root.service`, then the partition and filesystem are successfully resized.  I have yet to work out an automated approach though.

**Reply by rgilton** [reply]
Ok, I think I've figured out that `systemd-growfs-root.service` is just not configured by default to run on first-boot, or indeed on any boot.

So with the following modifications, the root fs gets resized properly for me:

 1. `/etc/repart.d/50-root.conf` contains the value I wrote in my previous comment.
 2. Set `systemd-growfs-root.service` to:
   a) by `WantedBy` by `multi-user.target`
   b) only run on first boot.
   c) be enabled

2(a) and (b) are achieved by this override in `/etc/systemd/system/systemd-growfs-root.service.d`:

```
[Unit]
ConditionFirstBoot=true

[Install]
WantedBy=multi-user.target
```

2(c) is achieved by running `systemctl enable systemd-growfs-root.service` in the `mkosi.finalize` script.

**Comment by fedya** [comment]
May you have any suggestions for me
i tried to grow my root partition with your configs
but it's still not working


```
/usr/lib/systemd/systemd-growfs /
Successfully resized "/" to 911.5M bytes.
```

```
systemctl status systemd-growfs-root.service
● systemd-growfs-root.service - Grow Root File System
     Loaded: loaded (/usr/lib/systemd/system/systemd-growfs-root.service; enabl>
     Active: active (exited) since Mon 2025-01-27 12:15:29 UTC; 2min 14s ago
 Invocation: a68a4218bf0a4b1088fb8defe3544b6c
       Docs: man:systemd-growfs-root.service(8)
    Process: 679 ExecStart=/usr/lib/systemd/systemd-growfs / (code=exited, stat>
   Main PID: 679 (code=exited, status=0/SUCCESS)
   Mem peak: 1.9M
        CPU: 91ms

Jan 27 12:15:29 elara systemd[1]: Starting Grow Root File System...
Jan 27 12:15:29 elara systemd-growfs[679]: Successfully resized "/" to 911.5M b>
Jan 27 12:15:29 elara systemd[1]: Finished Grow Root File System.
```

It's just 'grows' to the current size

**Comment by fedya** [comment]
i found what's wrong
forgot to add /etc/repart.d/50-root.conf
